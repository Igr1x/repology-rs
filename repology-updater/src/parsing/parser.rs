// SPDX-FileCopyrightText: Copyright 2025 Dmitry Marakasov <amdmi3@amdmi3.ru>
// SPDX-License-Identifier: GPL-3.0-or-later

use std::path::Path;

use crate::package::ParsedPackage;
use crate::parsing::error::PackageParsingError;
use crate::parsing::package_maker::PackageMaker;

pub trait PackageSink {
    fn push(&mut self, package_maker: PackageMaker) -> Result<(), PackageParsingError>;
}

pub trait PackageParser {
    fn parse(&self, path: &Path, sink: &mut dyn PackageSink) -> anyhow::Result<()>;
}

#[derive(Default)]
pub struct PackageAccumulator {
    pub packages: Vec<ParsedPackage>,
}

impl PackageSink for PackageAccumulator {
    fn push(&mut self, package_maker: PackageMaker) -> Result<(), PackageParsingError> {
        self.packages.push(package_maker.finalize()?);
        Ok(())
    }
}
