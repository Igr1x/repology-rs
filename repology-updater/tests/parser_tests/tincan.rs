// SPDX-FileCopyrightText: Copyright 2025 Dmitry Marakasov <amdmi3@amdmi3.ru>
// SPDX-License-Identifier: GPL-3.0-or-later

use std::path::Path;

use repology_updater::parsing::parser::PackageAccumulator;
use repology_updater::parsing::parser::PackageParser as _;
use repology_updater::parsing::parsers::tincan::TinCanParser;

#[test]
#[ignore]
fn test_tincan() {
    let mut packages = PackageAccumulator::default();
    TinCanParser {}
        .parse(Path::new("tests/parser_tests/tincan/ok"), &mut packages)
        .unwrap();
    insta::assert_debug_snapshot!(packages.packages);
}
