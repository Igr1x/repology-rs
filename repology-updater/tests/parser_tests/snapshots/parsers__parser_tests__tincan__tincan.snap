---
source: repology-updater/tests/parser_tests/tincan.rs
expression: packages.packages
snapshot_kind: text
---
[
    Package {
        srcname: "busybox",
        trackname: "busybox",
        visiblename: "busybox",
        projectname_seed: "busybox",
        rawversion: "1.36.1",
        maintainers: [
            "avs.origami@gmail.com",
        ],
        links: [
            Link {
                type: UpstreamDownload,
                url: "https://busybox.net/downloads/busybox-1.36.1.tar.bz2",
                fragment: None,
            },
        ],
        version: "1.36.1",
        flags: PackageFlags(
            0x0,
        ),
    },
]
